runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/config/StreamingGlobalConfigHandleImpl.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.dataflow.worker.streaming.config;
+
+import com.google.common.base.Preconditions;
+import java.util.Iterator;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import javax.annotation.Nonnull;
+import javax.annotation.concurrent.ThreadSafe;
+import org.apache.beam.sdk.annotations.Internal;
+
+@Internal
+@ThreadSafe
+public class StreamingGlobalConfigHandleImpl implements StreamingGlobalConfigHandle {
+
+  private final AtomicReference<StreamingGlobalConfig> streamingEngineConfig =
+      new AtomicReference<>();
+
+  private final CopyOnWriteArrayList<Consumer<StreamingGlobalConfig>> config_callbacks =
+      new CopyOnWriteArrayList<>();
+
+  @Override
+  public StreamingGlobalConfig getConfig() {
+    Preconditions.checkState(
+        streamingEngineConfig.get() != null,
+        "Global config should be set before any processing is done");
+    return streamingEngineConfig.get();
+  }
+
+  @Override
+  public void registerConfigObserver(@Nonnull Consumer<StreamingGlobalConfig> callback) {
+    StreamingGlobalConfig config;
+    synchronized (this) {
+      config_callbacks.add(callback);
+      config = streamingEngineConfig.get();
+    }
+    if (config != null) {
+      // read config from streamingEngineConfig again
+      // to prevent calling callback with stale config.
+      // The cached `config` will be stale if setConfig
+      // ran after the synchronized block.
+      callback.accept(streamingEngineConfig.get());
```

Comment:
on second thought this is still racy, we can call get() get the old result, pause before we call accept and then set config runs and accepts new result and then this accept runs.

Maybe we should call under some lock to prevent that? Or could have some single-threaded executor we use for running all the callbacks and here we schedule on that and then block for it to run?  Below in setConfig we could just schedule it to run and not have setConfig block on running the callbacks.

Reply:
Changed to use a single threaded executor.